clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -947,9 +947,15 @@ public String toString() {
     }
 
     public static class ValidString implements Validator {
-        final List<String> validStrings;
+        // visible for testing
```

Comment:
I am not a fan of "visible for testing". What about the following:

1. We add an enum to `StreamsConfig`:

```java
    public enum UpgradeFromValues {
        UPGRADE_FROM_0100("0.10.0"),
        ...
        UPGRADE_FROM_34("3.4");


        private final String value;

        UpgradeFromValues(final String value) {
            this.value = value;
        }

        public String toString() {
            return value;
        }
    }
```
2. We change the constants to:
```java
public static final String UPGRADE_FROM_34 = UpgradeFromValues.UPGRADE_FROM_34.toString();
```
3. We can use the enum in the tests and in the code.

Reply:
Since we just add a constant enum to StreamsConfig and given the hotfix character, I would say we do not need a KIP for this.

Reply:
This does not really help, because in the end we need a `List` of config values -- the source of truth what config value we support is inside the `CONFIG` object, so we need to get it from there IMHO.

If we don't want to tap into the `CONFIG` object, we can maintain a list ourselves, but I think it defeats the purpose of the test, as we again have a "split brain" problem for which we need to keep the `CONFIG` and our manually maintained list in sync.

Reply:
This part does not work: `public static final String UPGRADE_FROM_34 = UpgradeFromValues.UPGRADE_FROM_34.toString();` because we use
```
switch ((String) upgradeFrom) {
                case StreamsConfig.UPGRADE_FROM_0100:
                ...
}
```

And the compiler complain that `StreamsConfig.UPGRADE_FROM_0100` must be a const
```
error: constant string expression required
    case StreamsConfig.UPGRADE_FROM_0100:
```

Reply:
I pushed a corresponding update, but I am not super happy with it... I believe the original idea to get the list of values from `CONFIG` is safer. Curious to hear what others think. In the end I am fine both ways.

Reply:
You can use:
```java
switch (StreamsConfig.UpgradeFromValues.valueOf((String) upgradeFrom)) {
    case UPGRADE_FROM_0100:
    ...

```

Additionally, we can implement a validator that uses the enum to verify the values here:
```java
.define(UPGRADE_FROM_CONFIG,
                    ConfigDef.Type.STRING,
                    null,
                    in(null,
                       UPGRADE_FROM_0100,
                       UPGRADE_FROM_0101,
                       UPGRADE_FROM_0102,
                       UPGRADE_FROM_0110,
                       UPGRADE_FROM_10,
                       UPGRADE_FROM_11,
                       UPGRADE_FROM_20,
                       UPGRADE_FROM_21,
                       UPGRADE_FROM_22,
                       UPGRADE_FROM_23,
                       UPGRADE_FROM_24,
                       UPGRADE_FROM_25,
                       UPGRADE_FROM_26,
                       UPGRADE_FROM_27,
                       UPGRADE_FROM_28,
                       UPGRADE_FROM_30,
                       UPGRADE_FROM_31,
                       UPGRADE_FROM_32,
                       UPGRADE_FROM_33,
                       UPGRADE_FROM_34),
                    Importance.LOW,
                    UPGRADE_FROM_DOC)
```
So we do not need to maintain this list separately from the strings and enums.